src/uct/api/uct.h

```diff
@@ -1213,6 +1215,74 @@ struct uct_md_attr {
 };
 
 
+/**
+ * @ingroup UCT_MD
+ * @brief UCT MD memory attributes field mask
+ *
+ * The enumeration allows specifying which fields in @ref uct_md_mem_attr_t
+ * are present.
+ */
+enum uct_md_mem_attr_field {
+    UCT_MD_MEM_ATTR_FIELD_MEM_TYPE = UCS_BIT(0), /**< Indicate if memory type
+                                                      is populated. E.g. CPU/GPU */
+    UCT_MD_MEM_ATTR_FIELD_SYS_DEV  = UCS_BIT(1)  /**< Indicate if details of
+                                                      system device backing
+                                                      the pointer are populated.
+                                                      E.g. NUMA/GPU */
+};
+
+
+/**
+ * @ingroup UCT_MD
+ * @brief  Memory domain attributes.
+ *
+ * This structure defines the attributes of a memory pointer which may
+ * include the memory type of the pointer, and the system device that backs
+ * the pointer depending on the bit fields populated in field_mask.
+ */
+typedef struct uct_md_mem_attr {
+    /**
+     * Mask of valid fields in this structure, using bits from
+     * @ref uct_md_mem_attr_t. Note that the field mask is
+     * populated upon return from uct_md_mem_query and not set by user.
+     * Subsequent use of members of the structure are valid after ensuring that
+     * relevant bits in the field_mask are set.
+     */
+    uint64_t          field_mask;
+
+    /**
+     * The type of memory. E.g. CPU/GPU memory or some other valid type
+     */
+    ucs_memory_type_t mem_type;
+
+    /**
+     * Location of device pointer. E.g. NUMA/GPU
+     */
+    ucs_sys_device_t  sys_dev;
+} uct_md_mem_attr_t;
+
+
+/**
+ * @ingroup UCT_MD
+ * @brief Query attributes of a given pointer
+ *
+ * Return attributes such as memory type, and system device for the
+ * given pointer of specific length.
+ *
+ * @param [in]     md          Memory domain to run the query on. This function
+ *                             returns an error if the md does not recognize the
+ *                             pointer.
+ * @param [in]     address     The address of the pointer. Must be non-NULL.
+ * @param [in]     length      Length of the memory region to examine.
+ *                             Must be nonzero.
+ * @param [out]    mem_attr    If successful, filled with ptr attributes.
```

Comment:
what happens if zero is passed to the address or len ?

Reply:
@shamisp If either is true then `md_mem_query` should return an error. Is it better if that was explicitly stated? I can push that change if so.

Reply:
I would suggest to be more specific and define what error returned in this case. IMHO there are many code that don't have null check before calling function like this. Maybe instead of error we shell return status that the memory and device are undetected ?

Reply:
@shamisp Looking at `enum ucs_status_t` would returning UCS_ERR_UNSUPPORTED work or is this error reserved something else? If returning this error works, then I'll reference this for NULL/zero arguments.

Reply:
UCS_ERR_INVALID_PARAM sounds like more appropriate. Please documents this explicitly. thanks

Reply:
Does this look good?
```diff
diff --git a/src/uct/api/uct.h b/src/uct/api/uct.h
index bf59aadf1..816e39737 100644
--- a/src/uct/api/uct.h
+++ b/src/uct/api/uct.h
@@ -1272,9 +1272,11 @@ typedef struct uct_md_mem_attr {
  * @param [in]     md          Memory domain to run the query on. This function
  *                             returns an error if the md does not recognize the
  *                             pointer.
- * @param [in]     address     The address of the pointer. Must be non-NULL.
+ * @param [in]     address     The address of the pointer. Must be non-NULL
+ *                             else UCS_ERR_INVALID_PARAM error is returned.
  * @param [in]     length      Length of the memory region to examine.
- *                             Must be nonzero.
+ *                             Must be nonzero else UCS_ERR_INVALID_PARAM error
+ *                             is returned.
  * @param [out]    mem_attr    If successful, filled with ptr attributes.
  *
  * @return Error code.
```

Reply:
@shamisp If nothing else is pending, would you mind approving?